src/controllers/groupsController.js

```diff
@@ -0,0 +1,192 @@
+  }
+  if (memberType === "member") {
+    queryObject.members = { $in: [req.user.userID] };
+  }
+  if (!memberType) {
+    queryObject = {
+      $or: [{ owner: req.user.userID }, { members: req.user.userID }],
+    };
+  }
+
+  let result = await Group.find(queryObject)
+    .populate({
+      path: "owner",
+      select: "email name",
+      model: User,
+    })
+    .populate({
+      path: "members",
+      select: "email name",
+      model: User,
+    })
+    // .populate({
+    //   path: "events",
+    //   select: "_id name",
+    //   model: Event
+    // })
+    .exec(); //.populate("events");
+
+  return res.json({ count: result.length, groups: result });
+};
+
+// @desc    Endpoint for fetching one group
+// @route   GET /api/v1/groups/:id
+// @access  signed in users only
+const getGroup = async (req, res) => {
+  const { _id } = req.params;
+  const group = await Group.findOne({ _id })
+    .populate({
+      path: "owner",
+      select: "email name",
+      model: User,
+    })
+    .populate({
+      path: "members",
+      select: "email name",
+      model: User,
+    })
+    // .populate({
+    //   path: "events",
+    //   select: "_id name",
+    //   model: Event
+    // })
+    .exec();
+  if (!group) {
+    res.status(404);
+    throw new Error(`Group with ${_id} ID does not exist!`);
+  }
+
+  return res.json(group);
+};
+
+// @desc    Endpoint for creating a group
+// @route   POST /api/v1/activities
+// @access  signed in users only
+const createGroup = async (req, res) => {
+  const { groupName, description, memberEmails } = req.body;
+
+  // Find all users with matching emails
+  const foundUsers = await User.find(
+    { email: { $in: memberEmails } },
+    "_id email"
+  );
+
+  // Get all of the memberEmails that exist in the User database
+  const foundEmails = foundUsers?.map((user) => user.email);
+
+  // Get all of the attempted memberEmails that do not exist in the User database
+  const missingEmails = memberEmails?.filter(
+    (email) => !foundEmails.includes(email)
+  );
+
+  // Add the found userIDs to the group's members array
+  const memberIDs = foundUsers?.map((user) => user._id);
+
+  const newGroup = await Group.create({
+    groupName,
+    description,
+    owner: req.user.userID,
+    members: memberIDs,
+  });
+
+  let msg = "Group successfully created";
+  if (missingEmails?.length > 0) {
+    msg += ` but the following users could not be added: ${missingEmails}`;
+  }
+
+  // Sending back missing emails so FE can show who wasn't added (if necessary)
+  res.status(200).json({ msg, newGroup, missingEmails });
+};
+
+// @desc    Endpoint for updating a group
+// @route   PATCH /api/v1/groups/:id
+// @access  group owner only
+const updateGroup = async (req, res) => {
+  const { _id } = req.params;
+  const { groupName, description, memberEmails } = req.body;
+
+  if (groupName === "") {
+    res.status(400);
+    throw new Error("New group name must be provided!");
+  }
```

Comment:
My thinking was that if the groupName were included in the PATCH request, make sure the string isn't empty since it's required in the Group schema. I think you're right though in that it's probably much cleaner to change this to a PUT request and have the client provide a full copy of the group.  I'm sort of having trouble wrapping my head around how groupEvents are affected in a PUT request - since groupEvents are populated during GET requests, does this mean they don't need to be included in a Group PUT request?

Reply:
> My thinking was that if the groupName were included in the PATCH request, make sure the string isn't empty since it's required in the Group schema. 

Oh! Okay, that makes sense! But since there's validation on the group schema requiring that name be non-empty, we can rely on that validation to error when trying to save, and then the client will recieve the appropriate validation error message.

> I think you're right though in that it's probably much cleaner to change this to a PUT request and have the client provide a full copy of the group.

Another possibility is to only update the fields the user passes in. That's how description is working currently, but you'd have to make some adjustments to make sure that `memberEmails` is not overwritten with an empty array when a value is not passed for that property. Either option is a reasonable choice to me, but it would be best if the behavior is consistent for all params. (As I mentioned, currently if I don't include `memberEmails`, the data is overwritten with an empty array for the group -- my members are removed. You may be able to avoid this with just a check to see if `memberEmails` is `undefined` prior to setting `missingEmails` on line 145.)

Reply:
By the way, this is also something we could outline in a new task and do later as a refactor. Sometimes it makes sense to take on a little ["tech debt"](https://asana.com/resources/technical-debt) in order to prioritize shipping code faster (like when we're close to deadlines/ship dates), and we can come back and clean up when we have time.